### DISCUSSION TITLE
If neither whitelist nor blacklist is defined, the change is allowed (sort of fail-open behavior)

### DISCUSSION THREAD

**Discussion by mszabosimon-indeed** [argoproj/argo-cd#5331] [General]
Permissions for a project are defined by `clusterResourceWhitelist` or `clusterResourceBlacklist` and 
`namespaceResourceBlacklist` or `namespaceResourceWhitelist`. However I found that if neither whitelist nor blacklist  is defined, then all resources from that category (cluster or namespace) are allowed. 

This behavior seems to be consistent with the code found here: https://github.com/argoproj/argo-cd/blob/master/pkg/apis/application/v1alpha1/types.go#L2483-L2496 

Which brings me to the question: how to restrict a project to not make any changes to e.g. cluster resources? Looking at the code setting the `clusterResourceWhitelist` to an empty array won't work (as `len(proj.Spec.ClusterResourceWhitelist) == 0` will still be true).

**Comment by jopit** [comment]
I have not tried this, but what if you set `clusterResourceWhitelist` to something that does not exist?

**Reply by mszabosimon-indeed** [reply]
I believe that would work. Another workaround we found is to set the `clusterResourceBlacklist` to `"*"`. 

Meanwhile an issue has been filed about the same behavior: https://github.com/argoproj/argo-cd/issues/5540

**Reply by rumstead** [reply]
> I believe that would work. Another workaround we found is to set the `clusterResourceBlacklist` to `"*"`.
> 
> Meanwhile an issue has been filed about the same behavior: #5540

This is how we restrict cluster-level objects.  

Then we wanted to allow developers to create CRDs in some environments. We had to add the CRD to the allow list and left the deny list blank.

**Comment by emirot** [comment]
Just curious what is the default behavior when clusterResourceWhitelist or clusterResourceBlacklist when they are not defined?
It looks like it when neither are defined it will allow the creation of some resources but ClusterRole for instance won't get created.
Where can I find a list of resource that will and those that won't?
